Add more fortran wrapper tests

atztogo opened the pull request:
Motivation is to make each `test_fortran_SPG_FUNCTION_NAME.f90` contain multiple test functions (or subroutines) that use the same corresponding spglib function (or subroutine). An example is `test_fortran_spg_get_symmetry.f90`. Different set of test for different spglib function is written in another file (e.g., `test_fortran_spg_refine_cell.f90`).

This is not an elegant approach, but it is OK enough for me. If this way is reasonable, I will add more tests.

atztogo committed:
Add more fortran wrapper tests

atztogo committed:
Fix test_fortran_spg_refine_cell::test_FCC

LecrisUT commented on CMakeLists.txt:
```diff
@@ -61,6 +61,9 @@ if (spglib_WITH_Python)
 if (spglib_WITH_TESTS)
+    if(spglib_WITH_Fortran)
+        set(LIB_MOD_DIR ${CMAKE_CURRENT_BINARY_DIR}/fortran)
+    endif()
```
I would say use [`CMAKE_Fortran_MODULE_DIRECTORY`](https://cmake.org/cmake/help/latest/variable/CMAKE_Fortran_MODULE_DIRECTORY.html) instead. I.e., if it defined use that, otherwise define something. Will be enough to add `target_include_dir` to any `spglib_f08`. Will also need to fix the install for that.

atztogo replied:
Like this (https://github.com/spglib/spglib/pull/232/commits/a0dfab447faced40b70f02a9783b477e108110a9)?

LecrisUT replied:
Moved to https://github.com/spglib/spglib/pull/232/files#r1097094119

LecrisUT commented on test/fortran/CMakeLists.txt:
```diff
@@ -1,25 +1,15 @@
-# Add fortran tests here
-create_test_sourcelist(FortranTests_Files
-		${TestsDriver}
-		test_fortran_spg_get_symmetry.f90)
```
I would prefer to either use `create_test_sourcelist` or use an upstream Fortran unit tester like Fortuno.

This would minimize the executables displayed in IDEs, faster to build, and it is easier to add tests (less poilerplate compared to writing `Program` for each one)

atztogo replied:
I see your point.

I could not figure out how to put multiple test functions in one file, which is for easier writing and reading the test code. Using `create_test_sourcelist`, it registers one function per one file, as far as I tried. So if I want to add more tests, I thought I had to create a number of files. If this point is overcome, surely what you explained is better.

LecrisUT replied:
Totally agree with you. I would like it in a similar file as well.

Here is a design choice to consider though. Do we want `ctest` to be able to call subfunctions as well?

For example:
```cmake
# Loop over all main tests
foreach (test ${TestsToRun})
  get_filename_component(test_name ${test} NAME_WE)
  # Check if we have subtests defined
  if (${SubTests_${test_name}})
    # Loop over all subtests
    foreach (subtest ${SubTests_${test_name}}
      # Call test via ${FortranTests_Executable} ${test_name} ${subtest}
      # E.g.: ./FortranTests_Executable test_fortran_spg_get_symmetry test_rutile112
      add_test(NAME "${test_name}/${subtest}"
               COMMAND $<TARGET_FILE:FortranTests> ${test_name} ${subtest})
    endforeach ()
  else ()
    # Only run main test
    add_test(NAME "${test_name}"
             COMMAND $<TARGET_FILE:FortranTests> ${test_name})
  endif ()
endforeach ()
```

I can show you how to implement that from the `argc`/`argv` if you need help on that.

atztogo replied:
I see. It is a good trick. If I understand it correctly, I need two lists, `${TestsToRun}` and `${SubTests_${test_name}`. Am I correct?

> I can show you how to implement that from the argc/argv if you need help on that.

Yes, I need your help. I would appreciate it.

LecrisUT replied:
Ok, I'll push a commit directly here with an example for `test_fortran_spg_get_symmetry` maybe tonight or tomorrow.

LecrisUT commented on test/fortran/test_fortran_spg_get_symmetry.f90:
```diff
@@ -1,49 +1,98 @@
+    write (*, '("test_rutile112")')
+    call test_rutile112()
+    write (*, '("test_zincblende")')
+    call test_zincblende()
```
Note that using such bundling and a `stop`, failing one test will fail them all.

But also, this can be achieved with the function approach. There we can have as many function/subroutines we want in the file, as long as there is at least one function with the appropriate name and input/output format.

atztogo replied:
I understand this is a dirty approach. I could not find a better solution.

LecrisUT commented on test/fortran/test_fortran_spg_refine_cell.f90:
```diff
@@ -0,0 +1,33 @@
+
+        num_atom_bravais = spg_refine_cell(lattice, position, types, num_atom, symprec)
+
+        call assert_int(num_atom_bravais, 4)
```
Maybe worth making a:
```fortran
interface assert
  module procedure assert_int, assert_...
end interface
```

atztogo replied:
I agree! I am not yet familiar with fortran and just learning right now. Hopefully this will be done at near future.

LecrisUT replied:
Yeah Fortran is a mess of a language, I can barely get a footing into it without cursing

LecrisUT reviewed (commented):
What limitations were you trying to overcome here or how do you want to manage the tests?

atztogo committed:
Improve cmake for fortran wrapper test system

LecrisUT commented on CMakeLists.txt:
```diff
@@ -61,6 +61,9 @@ if (spglib_WITH_Python)
+        set(CMAKE_Fortran_MODULE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/fortran)
+    endif()
 	add_subdirectory(test)
 endif ()
```
Yeah, just add an `if (NOT)` statement so that devs can choose their own workflow, like this:

```suggestion
if (NOT CMAKE_Fortran_MODULE_DIRECTORY)
	set(CMAKE_Fortran_MODULE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/fortran_mods)
endif ()
```

Naming is arbitrary, that's what I implement in octopus.

As for the other fix in `spglib_f08`, I was refering to these lines:

https://github.com/atztogo/spglib/blob/a0dfab447faced40b70f02a9783b477e108110a9/fortran/CMakeLists.txt#L9-L11

The `target_include_directories` there is so that the Fortran compilers includes the folder where the `.mod` file is, either `CMAKE_Fortran_MODULE_DIRECTORY` during when we build locally (`BUILD_INTERFACE`), or the location where we installed it to (`CMAKE_INSTALL_INCLUDEDIR` at `INSTALL_INTERFACE`). So something like this:
```
target_include_directories(spglib_f08 PUBLIC
		"$<BUILD_INTERFACE:${CMAKE_Fortran_MODULE_DIRECTORY}>"
		"$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>")
```

One thing I didn't test, but it seemed to be working is that for Fortran modules (maybe even C++ modules in the future), we do not need to `include` the folder with the `.F90` file (as we do with `.h` in C/C++). In principle we could remove `spglib_f08.F90` from the install, but we need to test it out a bit further.

LecrisUT committed:
Soft revert to previous fortran tests

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
Add switchable subtests

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
Add useful labels for fortran tests

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT committed:
Change asserts to a common interface subroutine

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT commented:
Ok, a bit slow, but here is the implementation. I've added some labels to select the tests like `ctest -L fortran` or `ctest -L test_fortran_spg_get_symmetry`. I've tried to think if I can design a way to automatically parse the tests, but it was too troublesome eventually. Probably could do if I wrote a parser to list the tests, but then might as well move to Fortuno. Hopefully this is simple enough to maintain though.

LecrisUT committed:
Fix fortran linking for intel

Intel always links to its own MAIN__ definition. This disables it for `FortranTests` because MAIN__ is defined in  `fortran_test.c`

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT commented:
Fixes #234

atztogo commented:
Sorry I was a little bit busy not to work on this. @LecrisUT, to include your commits (I'm interested in how you add commits in this PR), can we merge this PR at this stage? This is because I am not good at handling github PR, and don't know how I can continue this PR to add more fix following your advice given two weeks ago. So merging, then making a new PR looks a reasonable way for me.

LecrisUT commented:
When yoy make a Github PR, you can set "Maintainers are allowed to edit this pull request.” that allows project collaborators to add commits to their (the person submitting the PR) fork. So you can continue to work on your branch, simply pull and merge/rebase your changes on top of it.

But, however it is more easier for you to work on. We can merge this and add a new branch/PR to continue to work on it.

atztogo commented:
Merged. Thanks @LecrisUT.
